packages/git/src/browser/git-widget.tsx

```diff
@@ -85,17 +85,17 @@ export class GitWidget extends GitDiffWidget implements StatefulWidget {
 
     @postConstruct()
     protected init() {
-        this.toDispose.push(this.repositoryProvider.onDidChangeRepository(repository =>
-            this.initialize(repository)
-        ));
         this.initialize(this.repositoryProvider.selectedRepository);
+        this.toDispose.push(this.repositoryProvider.onDidChangeRepository(newRepository => {
+            this.toDispose.dispose();
+            this.initialize(newRepository);
+        }));
         this.gitNodes = [];
         this.update();
     }
 
     async initialize(repository: Repository | undefined): Promise<void> {
         if (repository) {
-            this.toDispose.dispose();
```

Comment:
Could you just move this line up before if and revert all other changes?

Reply:
sorry, if i do that, ` this.toDispose.push(this.repositoryProvider.onDidChangeRepository(callback))` would be disposed when GitWidget is initialized.

Reply:
although changing a single line wouldn't work, but I can 
1) reverse the order of `this.initialize(...)` and `this.toDispose.push(this.repositoryProvider.onDidChangeRepository(callback))`, and 
2) move this line up

do you prefer this solution than the current change?

Reply:
now, I understand your struggle, sorry took a while :)

`this.toDispose` should be only used to dispose the entire widget, not on each initialize.

Could you create a new `DisposableCollection` for `initialize` method, something like:
```ts
protected readonly toDisposeOnInitialize = new DisposableCollection();
toDisposeOnInitialize().. {
   this.toDisposeOnInitialize.dispose();
   this.toDispose.push(this.toDisposeOnInitialize);
   ... // use toDisposeOnInitialize below not toDispose
   this.toDisposeOnInitialize.push(this.gitWatche...);
}
```

Reply:
i made the change that follows your design. please check if when you get the chance. Thank you for the review !